File: 46-permutations/permutations.java
class Solution {
    public void helper(int[] nums,List<Integer> ds,boolean[] isvalid,List<List<Integer>> ans){
        if(ds.size()==nums.length){
            List<Integer> list=new ArrayList<>();
            for(int i=0;i<ds.size();i++){
                list.add(ds.get(i));
            }
            ans.add(list);
            return;
        }for(int i=0;i<nums.length;i++){
            if(isvalid[i]==false){
                ds.add(nums[i]);
                isvalid[i]=true;
                helper(nums,ds,isvalid,ans);
                //backtracking
                isvalid[i]=false;
                ds.remove(ds.size()-1);
            }
        }
    }
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> ans=new ArrayList<>();    
        List<Integer> ds=new ArrayList<>();
        boolean[] isvalid=new boolean[nums.length]; //false
        helper(nums,ds,isvalid,ans);
        return ans;
    }
}

File: 46-permutations/permutations.cpp
#include <bits/stdc++.h>
using namespace std;  
class Solution {
public:
    void solve(int i,vector<int>& nums,vector<vector<int>> &perms){
        // Base case: if the last element is reached, add the permutation to perms
        int n=nums.size();
        if (i==n-1){
            perms.push_back(nums);
            return;
        }
        for(int j=i;j<n;j++){
            swap(nums[i],nums[j]);
            solve(i+1,nums,perms);
            swap(nums[i],nums[j]);//backtracking-restore original order
        }
    }
    vector<vector<int>> permute(vector<int>& nums) { 
        vector<vector<int>> perms;
        solve(0,nums,perms);
        return perms;
    } 
};